pkg/util/exithandler.go

```diff
@@ -41,14 +41,20 @@ func setupExitSignalHandling(doExit bool) {
 	go func() {
 		for sig := range sigChan {
 			log.Printf("captured %v, calling exit handlers and exiting..", sig)
-			exitRegistry.mutex.Lock()
-			for _, handler := range exitRegistry.handlers {
-				handler()
-			}
-			exitRegistry.mutex.Unlock()
+			ExecuteExitHandlers()
 			if doExit {
```

Comment:
When doExit is false, the exit handlers will run twice. It would be better to run them exactly once, either here, or in main().

Since the handlers are running under a mutex, we can clear the exitRegistry ensuring that handlers run exactly once.

Reply:
But I think we have a better way to handle this. The root cause is that we don't do graceful shutdown when receiving a signal, which is the expected behaviour. #284.

We can change the signal handler to run graceful shutdown (same as POST Stop). Then perform all exit handlers in main when runVfkit() returns. This will solve both issue with simpler code.

Reply:
> We can change the signal handler to run graceful shutdown (same as POST Stop). Then perform all exit handlers in main when runVfkit() returns. 

While I generally agree with what you wrote here and in #284 , there’s an additional complication with graceful shutdown: the guest may not shutdown (eg if there was a kernel panic). The corresponding API is https://developer.apple.com/documentation/virtualization/vzvirtualmachine/requeststop()?language=objc , the terminology it uses is to "request", "ask" the guest OS to stop.

If `runVfkit`  does not return after this request has been sent, we’ll have to decide if we force kill the VM or if we let that up to the user, in which case the cleanups won’t be done.

Reply:
> When doExit is false, the exit handlers will run twice. It would be better to run them exactly once, either here, or in main().
> 
> Since the handlers are running under a mutex, we can clear the exitRegistry ensuring that handlers run exactly once.

`doExit` is only false in the testing code, since I’d like to make a vfkit release, I’ll merge this as is, and this can be reworked as part of https://github.com/crc-org/vfkit/issues/284

Reply:
Sure, Stop may not happen, and this matches the semantics of SIGTERM, the process may handle it and never exit. This is typically enforced by SIGKILL after a timeout. So internally vfkit should probably use a timeout and fallback to HardStop, since we don't want to be killed without being able to clean up.